src/lib/translators/base.ts

```diff
@@ -0,0 +1,192 @@
+ * @param propertyKey the name of the method, in our case necessarily a step name
+ * @param descriptor the method descriptor
+ */
+function unsupported(target: BaseTranslator, propertyKey: S.PipelineStepName, descriptor: any) {
+  descriptor.value = function(step: S.PipelineStep) {
+    throw new StepNotSupported(step.name);
+  };
+  descriptor.value.__vqb_step_supported__ = false;
+}
+
+/**
+ * The `BaseTranslator` is the base class for all translators. It provides:
+ *
+ * - a default implementation for all pipeline steps that raises a
+ *   `StepNotSupported` error.
+ *
+ * - a `translate` entry point method that takes an array of `PipelineStep`
+ *   instances and return an array of objects matching the
+ *   `lib/matchers/OutputStep` type.
+ *
+ * - two properties `supportedSteps` and `unsupportedSteps` that respectively
+ *   return the list of supported and unssupported steps by the translator
+ *   class. These lists are computed automatically by instrospecting the
+ *   instance and testing, for each transformation step, if the corresponding
+ *   method holds a `__vqb_step_supported__` property. If this value is found
+ *   and set to `false`, then the step is considered as not supported.
+ *
+ * - a `supports` method that tests the support of a given step name.
+ */
+export class BaseTranslator implements StepMatcher<OutputStep> {
+  /**
+   * `supportedSteps` returns the list of steps supported by this translator class.
+   */
+  get supportedSteps(): Array<S.PipelineStepName> {
+    return transformerSteps()
+      .filter(stepname => this.constructor.prototype[stepname].__vqb_step_supported__ !== false)
+      .sort();
+  }
+
+  /**
+   * `unsupportedSteps` returns the list of steps _not_ supported by this translator class.
+   */
+  get unsupportedSteps(): Array<S.PipelineStepName> {
+    return transformerSteps()
+      .filter(stepname => this.constructor.prototype[stepname].__vqb_step_supported__ === false)
+      .sort();
+  }
+
+  /**
+   * `supports` returns a boolean indicating if the given step is supported or not.
+   *
+   * @param stepname the name of the step to test
+   */
+  supports(stepname: S.PipelineStepName) {
+    return !this.unsupportedSteps.includes(stepname);
+  }
+
+  @unsupported
+  filter(step: S.FilterStep) {}
+
+  @unsupported
+  domain(step: S.DomainStep) {}
+
+  @unsupported
+  select(step: S.SelectStep) {}
+
+  @unsupported
+  rename(step: S.RenameStep) {}
+
+  @unsupported
+  newcolumn(step: S.NewColumnStep) {}
+
+  @unsupported
+  delete(step: S.DeleteStep) {}
+
+  @unsupported
+  aggregate(step: S.AggregationStep) {}
+
+  @unsupported
+  custom(step: S.CustomStep) {}
+
+  /**
+   * translates an input pipeline into a list of steps that makes sense for the
+   * targeted backend.
+   *
+   * @param pipeline the array of input steps.
+   * @returns the list of translated output steps.
+   */
+  translate(pipeline: Array<S.PipelineStep>): Array<OutputStep> {
```

Comment:
in the case of a SQL translator, it would return a list of strings for example, that the caller would be have to join into a query string ?

Reply:
Probably, yes